lib/sycamore/sycamore/transforms/extract_graph.py

```diff
@@ -232,75 +149,190 @@ async def gather_api_calls():
-
-
-    -Instructions-
-    Entity Types and Descriptions: [{entities}]
-
-    Identify all entities that fit one of the following types and their descriptions.
-    For each of these entities extract the following information.
-    - entity_name: Name of the entity, capitalized
-    - entity_type: One of the following types listed above.
-
-    Format each entity that fits one of the types and their description as a json object.
-    Then, collect all json objects into a single json array named entities.
-
-    **Format Example:**
-    {{
-    entities: [
-    {{"name": <entity_name_1>, "type": <entity_type_1>}},
-    {{"name": <entity_name_2>, "type": <entity_type_2>}},
-    ...]
-    }}
+    Your job is to extract entities that match the entity schemas provided.
 
     -Real Data-
     ######################
-    Entity_types: {entities}
     Text: {query}
     ######################
     Output:"""
 
 
+class ResolveEntities:
+    """
+    Aggregates entities across all documents and resolves duplicate nodes created
+    during extraction.
+    """
+
+    @staticmethod
+    def resolve(docset: "DocSet") -> "DocSet":
+        docset.plan = ResolveEntities.GroupEntities(docset.plan)
+        docset = ResolveEntities._resolve(docset)
+        return docset
+
+    class GroupEntities(Map):
+        def __init__(self, child: Node, **resource_args):
+            super().__init__(child, f=self.GroupDocumentNodes, **resource_args)
+
+        @staticmethod
+        def GroupDocumentNodes(doc: HierarchicalDocument) -> HierarchicalDocument:
+            if "EXTRACTED_NODES" in doc.data:
+                return doc
+            nodes: defaultdict[dict, Any] = defaultdict(dict)
+            nodes |= doc["properties"].get("nodes", {})
+            for section in doc.children:
+                if "nodes" not in section["properties"]:
+                    continue
+                for label, hashes in section["properties"]["nodes"].items():
+                    for hash, node in hashes.items():
+                        if nodes[label].get(hash, None) is None:
+                            nodes[label][hash] = node
+                        else:
+                            for rel_uuid, rel in node["relationships"].items():
+                                nodes[label][hash]["relationships"][rel_uuid] = rel
+                del section["properties"]["nodes"]
+            doc["properties"]["nodes"] = nodes
+            return doc
+
+    @staticmethod
+    def _resolve(docset: "DocSet") -> "DocSet":
+        from sycamore import Execution
+        from sycamore.reader import DocSetReader
+        from ray.data.aggregate import AggregateFn
+
+        reader = DocSetReader(docset.context)
+
+        # Get list[Document] representation of docset, trigger execute with take_all()
+        execution = Execution(docset.context, docset.plan)
+        dataset = execution.execute(docset.plan)
+        docs = dataset.take_all()
```

Comment:
why not just use a map on the docset? or the dataset?

Reply:
Could do a map on the dataset, question is how do I add a document to a dataset? that's how I add the documents as nodes.

Reply:
Might be worth to change this in another PR when we make seperate transforms for entity/relationship resolution.